hal/src/common/dmac/transfer.rs

```diff
@@ -0,0 +1,857 @@
+        // operations beyond this fence.
+        // (see https://docs.rust-embedded.org/embedonomicon/dma.html#compiler-misoptimizations)
+        atomic::fence(atomic::Ordering::Acquire); // ▼
+
+        (self.config.source, self.config.dest, chan, self.payload)
+    }
+
+    /// Non-blocking; Immediately stop the DMA transfer and release all owned
+    /// resources
+    pub fn stop(self, dmac: &mut DmaController) -> (Src, Dst, Channel<Ready, ID>, P) {
+        let chan = self.chan.stop(dmac.as_mut());
+
+        // Memory barrier to prevent the compiler/CPU from re-ordering read/write
+        // operations beyond this fence.
+        // (see https://docs.rust-embedded.org/embedonomicon/dma.html#compiler-misoptimizations)
+        atomic::fence(atomic::Ordering::Acquire); // ▼
+
+        (self.config.source, self.config.dest, chan, self.payload)
+    }
+}
+
+/// These methods are available to a `DmaTransfer` holding a `Busy` channel
+impl<B, P, Src, Dst, const ID: u8> DmaTransfer<B, P, Src, Dst, Busy, ID>
+where
+    B: 'static + DmaBeat,
+    Src: DmaBuffer<B>,
+    Dst: DmaBuffer<B>,
+{
+    /// Issue a software trigger request to the corresponding channel.
+    /// Note that is not guaranteed that the trigger request will register,
+    /// if a trigger request is already pending for the channel.
+    #[inline(always)]
+    pub fn software_trigger(&mut self, dmac: &mut DmaController) {
+        self.xfer.software_trigger(dmac);
+    }
+    /// Blocking; Wait for the DMA transfer to complete and release all owned
+    /// resources
+    #[inline(always)]
+    pub fn wait(self, dmac: &mut DmaController) -> (Src, Dst, Channel<Ready, ID>, P) {
+        self.xfer.wait(dmac)
+    }
+
+    /// Non-blocking; Immediately stop the DMA transfer and release all owned
+    /// resources
+    #[inline(always)]
+    pub fn stop(self, dmac: &mut DmaController) -> (Src, Dst, Channel<Ready, ID>, P) {
+        self.xfer.stop(dmac)
+    }
+}
+
+/// Useable beat sizes for DMA transfers
+#[derive(Clone, Copy)]
+pub enum BeatSize {
+    /// Byte = [`u8`](core::u8)
+    Byte = 0x00,
+    /// Half word = [`u16`](core::u16)
+    HalfWord = 0x01,
+    /// Word = [`u32`](core::u32)
+    Word = 0x02,
+}
+
+// Trait enabling taking `*mut` references to references to a single `T`,
+/// references to arrays of `T` or slices of `T`, where `T: DmaBeat`,
+/// as well as required transfer length
+pub trait DmaBuffer<T: DmaBeat> {
+    /// Take a `*mut` reference to the last object
+    /// in the buffer needed by the DMAC. The pointer
+    /// changes depending on whether the address should
+    fn to_dma_ptr(&mut self) -> *mut T;
+    /// Check if the DMA address should be incrementing or not
+    fn incrementing(&self) -> bool;
+}
+
+impl<T: DmaBeat, const N: usize> DmaBuffer<T> for &mut [T; N] {
+    #[inline(always)]
+    fn to_dma_ptr(&mut self) -> *mut T {
+        let ptr = self.as_mut_ptr();
+        if self.incrementing() {
+            unsafe { ptr.add(N) }
```

Comment:
There's an easier way to do this. Use the `slice` method `as_mut_ptr_range`. It returns a `Range` struct with pointers to the first bye and the byte after the last. You can just ignore the `start` field and use the `end` field. It's easier and clearer than calculating it yourself.